### DISCUSSION TITLE
arm64 based Docker image?

### DISCUSSION THREAD

**Discussion by dasniko** [keycloak/keycloak#8846] [Archive]
Hi,

are there any plans to support/provide an arm64 based Docker image of Keycloak?
This would be awesome as it would improve not at least development with/for Keycloak on Mac M1 machines, but also other platforms offer arm based hardware.

\- Niko

[@stianst] Yes, we will deliver arm64 images for the new Quarkus distribution only, aiming to have this in Keycloak 18. https://github.com/keycloak/keycloak/issues/8825

**Comment by stianst** [comment]
I could see that as a possibility for new container images we are doing for Keycloak.X, but would depend on a number of factors like:

* Is the base image we're using available for arm64
* Can we build the images with GitHub Actions
* Can we test the images with GitHub Actions

If anyone from the community would like to do some research and propose something around this that would make it a lot more viable.

**Reply by dasniko** [reply]
Thanks, I understand.

The first requirement seems to be satisfiable: https://catalog.redhat.com/software/containers/ubi8/ubi-minimal/5c359a62bed8bd75a2c3fba8?architecture=arm64

GitHub Actions doesn't support arm64 runners yet, and also nothing on the roadmap about it.
This could only be accomplished with self-hosted runners, which I guess is not an option.

**Reply by stianst** [reply]
Yup, afraid so.

**Reply by tembleking** [reply]
Images can be built for ARM64 without too much effort: https://github.com/docker/build-push-action/blob/master/docs/advanced/multi-platform.md

**Reply by tembleking** [reply]
Maybe the support for ARM64 could be a best-effort, like other projects do, they provide an image for both AMD64 and ARM64, but they only test the AMD64 images until GitHub Actions provides ARM64 runners.

**Reply by zakjan** [reply]
Providing official ARM64 images (even untested) in [jboss/keycloak](https://hub.docker.com/r/jboss/keycloak) would be great for cross-platform development.

For now, I found [sleighzy/keycloak](https://hub.docker.com/r/sleighzy/keycloak) working well. It seems there is no change in build process needed.

**Reply by Teg79** [reply]
We need an official ARM64 image too for docker.

**Reply by ruckc** [reply]
@dasniko - i do quite a few amd64/arm64 container builds with qemu and docker buildx.  The workflow is very simple.  The only issue I see is that you don't want to run the tests in arm64 as the compute time is slow, but taking the quarkus tar.gz and building and arm64 container would take seconds.

I would submit a PR but I can't find the github workflow that builds the quarkus container... based on the quay.io/keycloak/keycloak it looks like a human is pushing them?

**Comment by istvandesign** [comment]
Hello, I'll bump this. I think ARM64 support should be a priority since a lot of us are switching to M1 Macs and this makes Keycloak a pain point in our setup and we need to build our own Keycloak image. Using third party images from other people increases the chance of a vulnerability.

**Comment by stianst** [comment]
Now that Quarkus distribution is the new supported distribution we're not going to consider ARM64 support for the legacy image, but should make this available to the new Quarkus images.

Due to the Docker licensing we're not able to use Docker to build Keycloak containers, but rather need to use Podman or Buildah. End result is the same though.

Right now we are building the containers manually with a script as a stop-gap solution as auto-builds stopped being a solution for us. We are planning on moving to GitHub Actions soon to replace this as we don't want and should not have manually steps in the release process.

We also need to consider how we will test the container on ARM64, which I'm not sure if there are any runners for ARM64 available on GitHub.

Was trying to build the [Quarkus container](https://github.com/keycloak/keycloak/tree/main/quarkus/container) for arm64, but it is failing for some reason:

```
$ podman build --platform linux/arm64 .
[1/2] STEP 1/8: FROM registry.access.redhat.com/ubi8-minimal AS build-env
Trying to pull registry.access.redhat.com/ubi8-minimal:latest...
Getting image source signatures
Checking if image destination supports signatures
Copying blob 5eb7814104c4 skipped: already exists  
Copying blob 02150aee4cb5 skipped: already exists  
Copying config b4f10be894 done  
Writing manifest to image destination
Storing signatures
[1/2] STEP 2/8: ENV KEYCLOAK_VERSION 18.0.0-SNAPSHOT
--> Using cache d2607d36e46c8e45102abfac4e7bdda8df839b8e03ab604a47b7b0cff301d00e
--> d2607d36e46
[1/2] STEP 3/8: ARG KEYCLOAK_DIST=https://github.com/keycloak/keycloak/releases/download/$KEYCLOAK_VERSION/keycloak-$KEYCLOAK_VERSION.tar.gz
--> Using cache e963d676b120de1ee00f1cec272932ef014519929658530662ada61f433cce0e
--> e963d676b12
[1/2] STEP 4/8: RUN microdnf install -y tar gzip
exec container process `/bin/sh`: Exec format error
```

**Reply by BuonOmo** [reply]
@stianst I'm not an expert and you've mentionned that Docker may not be a good fit, so really unsure about my comment. But have you seen [tembleking's comment](https://github.com/keycloak/keycloak/discussions/8846#discussioncomment-1975257)?

**Reply by Teg79** [reply]
@stianst using Buildah is a little more complex than --platform on docker buildx
Here you can find an example to build a multi-architecture image with Buildah: https://danmanners.com/posts/2022-01-buildah-multi-arch/

**Reply by ruckc** [reply]
@stianst - correct me if i'm wrong, but the docker licensing only applies to docker desktop... not docker provided github actions.  I use podman/buildah locally exclusively, but docker's github actions for handling the cross-architecture container builds.

**Reply by stianst** [reply]
Using Buildah/Podman will be fine, there's GitHub Actions for those as well. However, right now it seems an arm64 build doesn't work for some reason, would be good if someone could check if that's an issue with Docker for the new image: https://github.com/keycloak/keycloak/tree/main/quarkus/container

I looked at Docker's pricing, but couldn't really find out conclusively if it's only Desktop that is paid for. However, I feel much more comfortable with Podman/Buildah instead as those are more true open source projects. As developers we also need to be able to try/fix things locally with Podman.

Docker can also at any time change the pricing scheme for other things as well, like they did to Desktop with fairly short warning.

**Reply by stianst** [reply]
Figure out the solution to Podman build, just had to install qemu-user-static (https://lists.podman.io/archives/list/buildah@lists.buildah.io/thread/KKQSLFSVGSQI62EGFNG4Q3CTT23F7GUW/).

**Reply by BuonOmo** [reply]
Once again, not a Docker nor Keycloak expert. I've been using the script below to install keycloak (16.1.1) on my M1 Max:

```bash
set -e

if [[ "${KEYCLOAK_VERSION:-x}" = "x" ]]; then
    >&2 echo "$0: Variable KEYCLOAK_VERSION unset"
    exit 1
fi

kc_version=$KEYCLOAK_VERSION
kc_dir=$(mktemp -d)

trap "rm -rf $kc_dir" EXIT INT # Executed twice if CTRL+C.

git clone git@github.com:keycloak/keycloak-containers.git $kc_dir
cd $kc_dir/server

git checkout $kc_version
docker build -t "jboss/keycloak:${kc_version}" .
docker build -t "quay.io/keycloak/keycloak:${kc_version}" .
```

However, with keycloak 17, it fails:

```
[+] Building 33.1s (8/8) FINISHED                                                                                                                             
 => [internal] load build definition from Dockerfile                                                                                                     0.0s
 => => transferring dockerfile: 863B                                                                                                                     0.0s
 => [internal] load .dockerignore                                                                                                                        0.0s
 => => transferring context: 2B                                                                                                                          0.0s
 => [internal] load metadata for registry.access.redhat.com/ubi8-minimal:latest                                                                          1.8s
 => [1/4] FROM registry.access.redhat.com/ubi8-minimal@sha256:45b495856f6911192d93220bb773a881f51d52c0375c2a90574853c963d7a455                           4.2s
 => => resolve registry.access.redhat.com/ubi8-minimal@sha256:45b495856f6911192d93220bb773a881f51d52c0375c2a90574853c963d7a455                           0.0s
 => => sha256:45b495856f6911192d93220bb773a881f51d52c0375c2a90574853c963d7a455 1.47kB / 1.47kB                                                           0.0s
 => => sha256:9c8bf62c10dee0c0f55f01b40d4a084ad068e41f2625986476c02df38acd9380 737B / 737B                                                               0.0s
 => => sha256:b4f10be894d3f45e68346ebe536de9e883506e0c3073fce9f0536aa5ac074852 4.28kB / 4.28kB                                                           0.0s
 => => sha256:02150aee4cb58ca9f96e40db5f07c40cee540ab23baee823d4efe8f723f3135b 39.21MB / 39.21MB                                                         2.7s
 => => sha256:5eb7814104c45ac2cf2e4596b3704fc9745fc76dc214616274b4e2642dbc1ef9 1.74kB / 1.74kB                                                           0.7s
 => => extracting sha256:02150aee4cb58ca9f96e40db5f07c40cee540ab23baee823d4efe8f723f3135b                                                                1.3s
 => => extracting sha256:5eb7814104c45ac2cf2e4596b3704fc9745fc76dc214616274b4e2642dbc1ef9                                                                0.0s
 => [internal] load build context                                                                                                                        0.0s
 => => transferring context: 48.02kB                                                                                                                     0.0s
 => [2/4] RUN microdnf update -y && microdnf install -y glibc-langpack-en gzip hostname java-11-openjdk-headless openssl tar which && microdnf clean a  18.6s
 => [3/4] ADD tools /opt/jboss/tools                                                                                                                     0.0s
 => ERROR [4/4] RUN /opt/jboss/tools/build-keycloak.sh                                                                                                   8.3s
------
 > [4/4] RUN /opt/jboss/tools/build-keycloak.sh:
#8 0.146 Keycloak from [download]: https://github.com/keycloak/keycloak/releases/download/17.0.0/keycloak-17.0.0.tar.gz
#8 0.151   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
#8 0.152                                  Dload  Upload   Total   Spent    Left  Speed
100   659  100   659    0     0   2941      0 --:--:-- --:--:-- --:--:--  2941
100  155M  100  155M    0     0  21.9M      0  0:00:07  0:00:07 --:--:-- 24.2M
#8 7.256   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
#8 7.256                                  Dload  Upload   Total   Spent    Left  Speed
100 2333k  100 2333k    0     0  7291k      0 --:--:-- --:--:-- --:--:-- 7268k
#8 7.591   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
#8 7.591                                  Dload  Upload   Total   Spent    Left  Speed
100  806k  100  806k    0     0  3522k      0 --:--:-- --:--:-- --:--:-- 3522k
#8 7.831   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
#8 7.831                                  Dload  Upload   Total   Spent    Left  Speed
100  601k  100  601k    0     0  3135k      0 --:--:-- --:--:-- --:--:-- 3135k
#8 8.036   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
#8 8.036                                  Dload  Upload   Total   Spent    Left  Speed
100 1247k  100 1247k    0     0  5646k      0 --:--:-- --:--:-- --:--:-- 5646k
#8 8.263 /opt/jboss/tools/build-keycloak.sh: line 85: bin/jboss-cli.sh: No such file or directory
------
executor failed running [/bin/sh -c /opt/jboss/tools/build-keycloak.sh]: exit code: 127
```

**Reply by Teg79** [reply]
@BuonOmo Keycloak 17 is based on Quarkus, not on JBoss/WildFly, the script now is different

**Reply by ashim-kr-saha** [reply]
What is the new script?

**Comment by stianst** [comment]
We will provide ARM64 images for the new Quarkus distribution. Not quire sure when we are able to deliver it, and it may be blocked on updating to a new release process, but will try to get this done within the next couple months.

**Comment by benbeu** [comment]
This script here will build a local image for Keycloak 17.0.0, even for M1 Macs. Save this as an .sh file and execute it via your terminal. This is a good enough solution until the native port will be there. Right now Keycloak directly starts, but by disabling the last line this will just build the image.
Tested only for M1 chips and version 17.0.0

````
VERSION=17.0.0 # set version here
cd /tmp
git clone git@github.com:keycloak/keycloak.git
cd keycloak/quarkus/container
git checkout $VERSION
docker build -t "quarkus-keycloak:$VERSION" .
docker run -p 8080:8080 -e KEYCLOAK_ADMIN=admin -e KEYCLOAK_ADMIN_PASSWORD=admin "quarkus-keycloak:$VERSION" start-dev --http-relative-path /auth

**Comment by weberjoe** [comment]
This linux/**arm64** keycloak container solved it for me using **docker-compose**:

[wizzn/keycloak](https://hub.docker.com/r/wizzn/keycloak/tags)

Thats how it can be integrated into a .yaml file:

```
keycloak:
    image: wizzn/keycloak:12
    environment:
      - DB_VENDOR=POSTGRES
      - DB_ADDR=postgres
      - DB_PORT=5432
      - DB_DATABASE=keycloak
      - DB_USER=keycloak
      - DB_SCHEMA=public
      - DB_PASSWORD=password
      - KEYCLOAK_USER=admin
      - KEYCLOAK_PASSWORD=secure
    ports:
      - "8080:8080"
..
```

**Comment by stianst** [comment]
We're now providing multi-arch for Quarkus and WildFly distributions

**Reply by SimonMacIntyre** [reply]
> We're now providing multi-arch for Quarkus and WildFly distributions

For anyone wondering this appears to be for 17+ only, so 17.0.1-legacy if you are on wildfly. We were on 16.1.1 and went to 17.0.1-legacy and it seemed to work. Building 16.1.1 locally on the m1 that did work for others, did not work seamlessly for us.

**Reply by dimitarvdimitrov** [reply]
excuse me for the ignorance but can you provide a link?

**Reply by kosssi** [reply]
https://quay.io/repository/keycloak/keycloak?tab=tags ;)
